raft/internal.hh

```diff
@@ -10,12 +10,18 @@
 #include <ostream>
 #include <functional>
 #include "utils/UUID.hh"
+#include "utils/tagged_integer.hh"
 
 namespace raft {
 namespace internal {
 
+// Unlike utils::tagged_uuid, tagged_id has a non-final idl type.
 template<typename Tag>
 using tagged_id = utils::tagged_uuid<Tag>;
 
+// Unlike utils::tagged_integer, tagged_uint64 has a non-final idl type.
+template<typename Tag>
+using tagged_uint64 = utils::tagged_integer<Tag, uint64_t>;
```

Comment:
Did you check that given serializer for `tagged_uint64` and  `utils::tagged_integer` right overload is chosen. The one is the alias for another.

Reply:
Yes.
Although reviewing the generated code is not much helpful since the compiler needs to choose the serializer overload as you noted, I tried the following:
```
@@ -22,7 +22,7 @@ struct tagged_id {
 
 template<typename Tag>
 struct tagged_uint64 {
-    uint64_t value();
+    uint64_t get_value();
 };
 
 } // namespace internal
```

And saw that the compiler emits error for the expected types:
```
build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::term_tag, unsigned long>'
build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::index_tag, unsigned long>'
build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::term_tag, unsigned long>'
```

Reply:
and the generated serializer also shows that tagged_uint64 is not final:
```c++
template <typename Tag>
template <typename Output>
void serializer<raft::internal::tagged_uint64<Tag>>::write(Output& buf, const raft::internal::tagged_uint64<Tag>& obj) {
  set_size(buf, obj);
  static_assert(is_equivalent<decltype(obj.get_value()), uint64_t>::value, "member value has a wrong type");
  serialize(buf, obj.get_value());
}

template <typename Tag>
template <typename Input>
raft::internal::tagged_uint64<Tag> serializer<raft::internal::tagged_uint64<Tag>>::read(Input& buf) {
 return seastar::with_serialized_stream(buf, [] (auto& buf) {
  size_type size = deserialize(buf, boost::type<size_type>());
  auto in = buf.read_substream(size - sizeof(size_type));
  auto __local_0 = deserialize(in, boost::type<uint64_t>());

  raft::internal::tagged_uint64<Tag> res {std::move(__local_0)};
  return res;
 });
}
```

Reply:
> Yes. Although reviewing the generated code is not much helpful since the compiler needs to choose the serializer overload as you noted, I tried the following:
> 
> ```
> @@ -22,7 +22,7 @@ struct tagged_id {
>  
>  template<typename Tag>
>  struct tagged_uint64 {
> -    uint64_t value();
> +    uint64_t get_value();
>  };
>  
>  } // namespace internal
> ```
> 
> And saw that the compiler emits error for the expected types:
> 
> ```
> build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::term_tag, unsigned long>'
> build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::index_tag, unsigned long>'
> build/dev/gen/idl/raft_storage.dist.impl.hh:53:44: error: no member named 'get_value' in 'utils::tagged_integer<raft::term_tag, unsigned long>'
> ```

I am still skeptical that it will always work. May be it works just because you do not have any other tagged integer with uint64_t type. I do not see why it would always work because those are not to different typed in the type system.

Reply:
It works exactly like `tagged_id` above it that is an alias of tagged_uuid.
And the latter has no ValueType, so apparently the compiler does distinguish between correctly and probably infers the "explicit" overload over the "implicit" one.

Reply:
I believe that tagged_uint64_t and tagged_id can be defined as derived class of tagged_integer and tagged_uuid, respectively, but other than paranoia, I'm not sure the're a good reason to do so.

Reply:
> It works exactly like `tagged_id` above it that is an alias of tagged_uuid.

tagged_uuid is not in the idl, so there is no chance a compiler will choose incorrect serializer.

> I believe that tagged_uint64_t and tagged_id can be defined as derived class of tagged_integer and tagged_uuid, respectively, but other than paranoia, I'm not sure the're a good reason to do so.

I am afraid that if some other code will define `utils::tagged_integer<Tag, uint64_t>` by some other name a compiler may choose wrong serizlier depending on an order of includes. I do not see any evidence that this is impossible.

Reply:
@gleb-cloudius it's not a matter of include order (maybe it used to be in the distant past).
The compiler should choose the override best matching the type.

See https://godbolt.org/z/jsqeob8fh

Both clang and gcc choose the right overload regardless of order.

Reply:
> tagged_uuid is not in the idl, so there is no chance a compiler will choose incorrect serializer.

True.

We explicitly defined types for the types that use it, like:
```c++
class table_id final {
    utils::UUID uuid();
};
```

We can do the same for the tagged_uint64_t types like raft::term_t and friends.

Reply:
> > tagged_uuid is not in the idl, so there is no chance a compiler will choose incorrect serializer.
> 
> True.
> 
> We explicitly defined types for the types that use it, like:
> 
> ```c++
> class table_id final {
>     utils::UUID uuid();
> };
> ```
> 
> We can do the same for the tagged_uint64 types like raft::term_t and friends.

Lets do it. I will sleep better.